[SPARK-34544][PYTHON] Convert PandasDataFrameLike and PandasSeriesLike to aliases of Pandas types

zero323 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This PR proposes replacing currently used `Protocols`:

- `PandasDataFrameLike`
- `PandasSeriesLike`

with simple aliases of upstream types.

This exposed a number of typing issues, pirmairly around `pyspark.pandas` API, which will be resolved in this PR.


Additionally it adds [VirtusLab/pandas-stubs](https://github.com/VirtusLab/pandas-stubs) to CI dependencies.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Currently used `Protocols` where a workaround, included to improve typing coverage until Pandas exposes their type hints.

In a meantime, relatively stable stubs sources emerged (with a lot of ongoing discussions around) and `pandas-stubs` package (available on PyPI and conda-forge) provides better coverage without adding maintenance overhead on our side.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

Better typing experience around Pandas UDFs.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

Existing typecheck pipeline and unit tests.

zero323 committed:
Replace Protocols with aliases and add basic overloads

These includes:
- Basic casts to the expected output types
- Ignores on `pandas.api.type` and `pandas.plotting._matplotlib` imports ‒
  in general, this import objects not in `__all__` and could be replaced
  with more precise imports (i.e. from `from pandas.core.dtypes.common`
  for the former)

zero323 committed:
Resolve type issues in pandas.frame.DataFrame.{first, last}

zero323 committed:
Use bound var for ColumnOrName

zero323 committed:
Adjust return annotation for _get_or_create_repr_pandas_cache

When called in Series.__repr__

https://github.com/apache/spark/blob/6e45b04db48008fa033b09df983d3bd1c4f790ea/python/pyspark/pandas/series.py#L6361-L6367

```python
    def __repr__(self) -> str_type:
        max_display_count = get_option("display.max_rows")
        if max_display_count is None:
            return self._to_internal_pandas().to_string(name=self.name, dtype=self.dtype)

        pser = self._psdf._get_or_create_repr_pandas_cache(max_display_count)[self.name]
        pser_length = len(pser)
```

it actually retrieves a pandas Series object.

zero323 committed:
cast values in _select_rows_by_slice

Seems like mypy has problems with infering types here.

zero323 committed:
Add ignore on _cleanup_and_return

Override violates Liskov substitution principle

zero323 committed:
Ignore incompatible overrides in generic.Frame.truncate

zero323 committed:
add pandas stubs to github workflow

zero323 commented:
FYI @HyukjinKwon 

There is still a lot of work to done here (reduced typing errors from ~150  to  <50 so far).

SparkQA commented:
**[Test build #146302 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146302/testReport)** for PR 34927 at commit [`5958791`](https://github.com/apache/spark/commit/5958791957c391fcd63e77b9be174ccb255887e5).

SparkQA commented:
**[Test build #146302 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146302/testReport)** for PR 34927 at commit [`5958791`](https://github.com/apache/spark/commit/5958791957c391fcd63e77b9be174ccb255887e5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

HyukjinKwon commented:
I like this!!

HyukjinKwon commented:
cc @ueshin @itholic @xinrong-databricks FYI

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50774/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50774/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146302/

itholic commented:
A brief review of the work so far looks good to me!!

Let me revisit and take a closer look when it's ready to review.

zero323 committed:
Restrict dtype argument type for pandas.data_type_ops _as_*

These methods call

```
codes._internal.data_fields[0].copy
```

where `codes._internal.data_fields[0]` is `InternalField` which copy
expects Union[Dtype, _NoValueType]

https://github.com/apache/spark/blob/6fb417e21c98283c40713698c58a73e68ea9614f/python/pyspark/pandas/internal.py#L171-L179

zero323 committed:
Cast name argument for StructField to str

Pandas names can be any Hashable objects

zero323 committed:
Ignore asignment to val

zero323 committed:
Ignore SPARK-37668

zero323 committed:
Cast results of _to_pandas conversions

zero323 committed:
Ignore GroupBy._make_pandas_df_builder_func func arg-type

We have `Callable[[Any], Union[DataFrame, Series]]` but expect `Callable[[DataFrame], DataFrame]`
‒ this seems to be accounted for in the control flow.

zero323 committed:
Ignore first_series arg type

Accounted for in control flow

zero323 committed:
Ignore access to internal _values and dtype type mismatch

zero323 committed:
Ignore imprecise axis usage

`Axis` should be defined as

```python
Union["Literal[0]", "Literal[1]", "Literal['index']", "Literal['columns']"]
```

but this change escalates all over the place, so let's handle it
separately.

zero323 committed:
Ignore unsupported operand type for DateOffset arithmetics

This seems to be limitation of the external annotations.

zero323 committed:
Ignoring slicing of datet time index with datettime

zero323 committed:
Ignore getitem on index

Seems like @cache_readonly is not recognized as "property like"
decorator and index is inferred as nullary method.

SparkQA commented:
**[Test build #146332 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146332/testReport)** for PR 34927 at commit [`3c5c894`](https://github.com/apache/spark/commit/3c5c89408ffa02e726657c25a880b4ae2de7f528).

SparkQA commented:
**[Test build #146332 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146332/testReport)** for PR 34927 at commit [`3c5c894`](https://github.com/apache/spark/commit/3c5c89408ffa02e726657c25a880b4ae2de7f528).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50807/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146332/

zero323 committed:
Ignore test mismatches

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50807/

zero323 committed:
Ignore droplevel argument type

Upstream annotations are inconsistent with description

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50807/

zero323 committed:
Cast dtype in astype method

Restricted by control flow

zero323 committed:
Allow passing Pandas-on-Spark df to pd.DataFrame.info

zero323 committed:
Cast first_series to the expected component of Union

zero323 committed:
Ignore incompatibilities in from_dict

We could synch types once pandas annotations mature enough

zero323 committed:
Cast to DataFrame in DataFrame.__repr__

zero323 committed:
Resolve typing issues around Series.spark

zero323 committed:
Ignore np.issubdtype dtype type

It seems like ExtensionDtype cannot be used there,

zero323 committed:
Cast .quantile result to DataFrame

Based on control flow. Could be overloaded in the method

zero323 committed:
Cast partitionBy argument

zero323 committed:
Cast ambigous first_series arguments

Based on control flow

zero323 committed:
Convert name and dtype to booleans before calling pandas to_string

https://github.com/pandas-dev/pandas/blob/a3702e2207af106939804ceb193bf4fae3e23a31/pandas/core/series.py#L1534-L1537

zero323 committed:
Ignore columns reasign

zero323 committed:
Ignore on SPARK-37678

zero323 committed:
Type ignore Index.spark

Type cannot be more precise, because hole param is unbound

zero323 committed:
Ignore parse_dates incompatibility

Issue with upstream annotations

zero323 committed:
Ignore columns reasign

zero323 committed:
Ignore pandas __class_getitem__ assignments

zero323 committed:
Ignore _all_kinds access

Internal, not exposed in upstream hints

SparkQA commented:
**[Test build #146339 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146339/testReport)** for PR 34927 at commit [`e5043a5`](https://github.com/apache/spark/commit/e5043a5c0d3d13d3989aac39fc572919b372e58a).

SparkQA commented:
**[Test build #146339 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146339/testReport)** for PR 34927 at commit [`e5043a5`](https://github.com/apache/spark/commit/e5043a5c0d3d13d3989aac39fc572919b372e58a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50813/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146339/

zero323 committed:
Drop Pandas Protocols tests

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50813/

zero323 commented:
> I've brief reviewed of the work, and so far looks pretty nice to me!!
> 
> Let me revisit and take a closer look when it's ready to review.

I'll ping you once I am closer to completion ‒ I resolved most of the minor problems, but hit a bigger issue with `pandas_udf` annotations ‒ seems like we'll have to rethink how this stuff is done, which is something I am not very happy about.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50813/

SparkQA commented:
**[Test build #146341 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146341/testReport)** for PR 34927 at commit [`d81bfee`](https://github.com/apache/spark/commit/d81bfee6b56ffffa3deac2ab14e8a88821998e77).

SparkQA commented:
**[Test build #146341 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146341/testReport)** for PR 34927 at commit [`d81bfee`](https://github.com/apache/spark/commit/d81bfee6b56ffffa3deac2ab14e8a88821998e77).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

ueshin commented:
We can remove the pandas entry from `mypy.ini`?
https://github.com/apache/spark/blob/d81bfee6b56ffffa3deac2ab14e8a88821998e77/python/mypy.ini#L119-L120

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50815/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146341/

zero323 committed:
Fix udf annotations and examples

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50815/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50815/

zero323 committed:
Drop pandas ignores from mypy.ini

SparkQA commented:
**[Test build #146356 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146356/testReport)** for PR 34927 at commit [`045d587`](https://github.com/apache/spark/commit/045d58728c3873838eb318592869fc4840137815).

zero323 committed:
Add short description of purpose to pandas protocols

SparkQA commented:
**[Test build #146357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146357/testReport)** for PR 34927 at commit [`111b6cc`](https://github.com/apache/spark/commit/111b6ccae469b16d4e21389917d8b6e162be3cb8).

zero323 committed:
Drop ignore[import] on pandas imports

SparkQA commented:
**[Test build #146358 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146358/testReport)** for PR 34927 at commit [`535e3fb`](https://github.com/apache/spark/commit/535e3fbc821bdc53b27a2eabb172923a87d614f9).

SparkQA commented:
**[Test build #146356 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146356/testReport)** for PR 34927 at commit [`045d587`](https://github.com/apache/spark/commit/045d58728c3873838eb318592869fc4840137815).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #146357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146357/testReport)** for PR 34927 at commit [`111b6cc`](https://github.com/apache/spark/commit/111b6ccae469b16d4e21389917d8b6e162be3cb8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50830/

SparkQA commented:
**[Test build #146358 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146358/testReport)** for PR 34927 at commit [`535e3fb`](https://github.com/apache/spark/commit/535e3fbc821bdc53b27a2eabb172923a87d614f9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50831/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50832/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146358/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146357/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146356/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50830/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50831/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50832/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50832/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50831/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50830/

SparkQA commented:
**[Test build #146362 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146362/testReport)** for PR 34927 at commit [`28684cd`](https://github.com/apache/spark/commit/28684cda22b8e8c4e5ea0292899d28c221bcaa84).

SparkQA commented:
**[Test build #146362 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146362/testReport)** for PR 34927 at commit [`28684cd`](https://github.com/apache/spark/commit/28684cda22b8e8c4e5ea0292899d28c221bcaa84).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50836/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146362/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50836/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50836/

SparkQA commented:
**[Test build #146370 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146370/testReport)** for PR 34927 at commit [`47a727d`](https://github.com/apache/spark/commit/47a727d381768706167ba1ae19eeca8b4978d9a5).

SparkQA commented:
**[Test build #146370 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146370/testReport)** for PR 34927 at commit [`47a727d`](https://github.com/apache/spark/commit/47a727d381768706167ba1ae19eeca8b4978d9a5).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146370/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50844/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50844/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50844/

zero323 committed:
Merge branch 'master' into SPARK-34544

SparkQA commented:
**[Test build #146445 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146445/testReport)** for PR 34927 at commit [`1224d1e`](https://github.com/apache/spark/commit/1224d1e99a781c11c5c53011af4272c40fa255a7).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50920/

SparkQA commented:
**[Test build #146445 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/146445/testReport)** for PR 34927 at commit [`1224d1e`](https://github.com/apache/spark/commit/1224d1e99a781c11c5c53011af4272c40fa255a7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait ExposesMetadataColumns extends LogicalPlan `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/146445/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/50920/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/50920/

itholic commented on .github/workflows/build_and_test.yml:
```diff
@@ -478,6 +478,7 @@ jobs:
         # Jinja2 3.0.0+ causes error when building with Sphinx.
         #   See also https://issues.apache.org/jira/browse/SPARK-35375.
         python3.9 -m pip install 'flake8==3.9.0' pydata_sphinx_theme 'mypy==0.920' numpydoc 'jinja2<3.0.0' 'black==21.5b2'
+        python3.9 -m pip install pandas-stubs
```
just qq: Why new line instead of just adding to the existing list?

`python3.9 -m pip install 'flake8==3.9.0' pydata_sphinx_theme 'mypy==0.920' numpydoc 'jinja2<3.0.0' 'black==21.5b2' pandas-stubs`

itholic replied:
Oh..the above line is for build. never mind!

itholic commented on python/pyspark/pandas/frame.py:
```diff
@@ -3989,7 +3989,7 @@ def insert(
 
         if is_name_like_tuple(column):
-            if len(column) != len(self.columns.levels):
+            if len(column) != len(self.columns.levels):  # type: ignore[attr-defined]  # SPARK-37668
```
nit: haven't strong feeling, but I think it would be better if we have comment about what is `SPARK-37668` for.

itholic reviewed (approved):
Thanks for the bunch of cleanup!

zero323 commented:
Merged into master.

Thanks all!
